File: engine/operations/executor.py
"""已审批操作的原生写入执行。"""
from __future__ import annotations

from collections.abc import Callable

from ..context import EngineContext
from ..errors import (
    AgentReferenceError,
    AgentRequestError,
    ConcurrentModificationError,
    require_agent_capability,
)
from ..sessions import agent_read
from ..sessions.index import AgentSessionIndex
from . import metadata, verification as probe_mod
from .delete import SessionDeletionService
from .metadata_store import metadata_key
from .edit import EditOperationHandler
from .migrate import MigrationService
from .plan_store import OperationPlan


class OperationExecutor:
    def __init__(
        self,
        ports: EngineContext,
        index: AgentSessionIndex,
        migration: MigrationService,
        edit: EditOperationHandler,
        database: Callable,
    ):
        self._ports = ports
        self._index = index
        self._migration = migration
        self._edit = edit
        self._database = database

    def execute(self, operation: OperationPlan) -> dict:
        handlers = {
            "edit": self._apply_edit,
            "migration": self._apply_migration,
            "metadata": self._apply_metadata,
            "delete": self._apply_delete,
        }
        handler = handlers.get(operation.kind)
        if handler is None:
            raise AgentRequestError(
                "operation kind 非法", {"kind": operation.kind},
            )
        return handler(operation)

    def _apply_edit(self, operation: OperationPlan) -> dict:
        params = operation.input()
        if params["probe"]:
            require_agent_capability(
                self._ports.adapter(params["tool"]), "probe", "verifier",
            )
        return self._edit.apply(operation, self._finish_mutation)

    def _finish_mutation(self, tool, editor, result, document, snapshot, probe):
        if not probe:
            return result
        try:
            verifier = require_agent_capability(
                self._ports.adapter(tool), "probe", "verifier",
            )
            report = verifier.probe_edited(
                editor, document, result,
            )
        except probe_mod.ProbeTimeout as error:
            report = probe_mod.timeout_report(tool, error)
        result["probe"] = report
        if report["status"] == "passed":
            return result
        if snapshot:
            editor.restore_snapshot(snapshot, document)
            result.update(ok=False, error="隔离探针未通过,已自动还原快照")
        return result

    def _apply_migration(self, operation: OperationPlan) -> dict:
        params = operation.input()
        try:
            record = self._index.resolve(
                params["source_tool"], params["ref"],
            )
        except AgentReferenceError as error:
            raise ConcurrentModificationError(
                "会话在迁移计划生成后已变化，请重新计划"
            ) from error
        if record.revision != operation.base_revision:
            raise ConcurrentModificationError(
                "会话在迁移计划生成后已变化，请重新计划"
            )
        session = agent_read.read_indexed_session(self._index, record)
        result = self._migration.apply(
            params["source_tool"],
            params["target_tool"],
            record.canonical_ref,
            probe=params["probe"],
            max_turn=params.get("max_turn"),
            probe_model=params.get("probe_model"),
            session=session,
        )
        structure = result.get("validation", {}).get("structure", {})
        if result.get("rolled_back") or structure.get("ok") is not True:
            raise RuntimeError("迁移写入后的结构校验失败，产物已回滚")
        return result

    def _apply_metadata(self, operation: OperationPlan) -> dict:
        params = operation.input()
        try:
            record = self._index.resolve(params["tool"], params["ref"])
        except AgentReferenceError as error:
            raise ConcurrentModificationError(
                "会话在元数据计划生成后已变化，请重新计划"
            ) from error
        if record.revision != operation.base_revision:
            raise ConcurrentModificationError(
                "会话在元数据计划生成后已变化，请重新计划"
            )
        if record.row.get("id") != params["session_id"]:
            raise ConcurrentModificationError(
                "会话标识在元数据计划生成后已变化，请重新计划"
            )
        result = metadata.compare_and_set_entry(
            params["tool"],
            params["session_id"],
            params["metadata_before"],
            params["patch"],
            self._ports,
        )
        return {"metadata": result}

    @staticmethod
    def _protected_cause(metadata_row: dict) -> str | None:
        if metadata_row.get("pinned") is True:
            return "pinned"
        if metadata_row.get("archived") is True:
            return "archived"
        if metadata_row.get("tags"):
            return "tagged"
        return None

    def _apply_delete(self, operation: OperationPlan) -> dict:
        params = operation.input()
        tool = params["tool"]
        # 计划期的保护审查挡不住批准前才被 pin/archive/打标签的会话:元数据
        # 与会话内容 revision 无关,逐条 revision 比对不可能发现这种变化。
        metadata_rows = metadata.list_all(self._ports)
        deletion = SessionDeletionService(self._ports)
        result = {"succeeded": [], "skipped": [], "failed": []}
        for target in params["targets"]:
            ref = target["ref"]
            try:
                protected = self._protected_cause(
                    metadata_rows.get(
                        metadata_key(tool, target["session_id"]), {},
                    ),
                )
                if protected is not None:
                    result["skipped"].append({
                        "tool": tool,
                        "ref": ref,
                        "cause": "protected",
                        "protection": protected,
                    })
                    continue
                try:
                    record = self._index.resolve(tool, ref)
                except AgentReferenceError as error:
                    reason = error.params.get("reason")
                    result["skipped"].append({
                        "tool": tool,
                        "ref": ref,
                        "cause": "changed" if reason == "session_changed" else "not_found",
                    })
                    continue
                if (
                    record.row.get("id") != target["session_id"]
                    or record.revision != target["revision"]
                ):
                    result["skipped"].append({
                        "tool": tool,
                        "ref": ref,
                        "cause": "changed",
                    })
                    continue
                deletion.delete(tool, record.canonical_ref)
                self._index.evict(tool, record.canonical_ref)
                result["succeeded"].append({"tool": tool, "ref": ref})
            except Exception as error:  # noqa: BLE001 - 单条失败继续批处理
                result["failed"].append({
                    "tool": tool,
                    "ref": ref,
                    "error": str(error)[:500],
                })
        return result

File: crates/ferry-engine/src/operations/executor.rs
//! 已审批操作的原生写入执行。
//!
//! 硬约束：
//! - 每条分支都要在写之前重新解析索引并比 `base_revision`（§2.4）；
//! - migration 的二次门禁：`rolled_back or structure.ok is not True → RuntimeError`（第 25 条）。

use serde_json::{Map, Value};

use crate::errors::DomainError;
use crate::operations::edit::EditOperationHandler;
use crate::operations::metadata;
use crate::operations::migrate::MigrationService;
use crate::operations::plan_store::OperationPlan;
use crate::operations::types::{EngineError, EngineResult, Ports, Resolver};

pub struct OperationExecutor {
    ports: Ports,
    index: Resolver,
    migration: MigrationService,
    edit: EditOperationHandler,
}

impl OperationExecutor {
    pub fn new(ports: Ports, index: Resolver) -> Self {
        Self {
            migration: MigrationService::new(Ports::clone(&ports)),
            edit: EditOperationHandler::new(Ports::clone(&ports), Resolver::clone(&index)),
            ports,
            index,
        }
    }

    pub fn execute(&self, operation: &OperationPlan) -> EngineResult<Value> {
        match operation.kind.as_str() {
            "edit" => self.apply_edit(operation).map(Value::Object),
            "migration" => self.apply_migration(operation).map(Value::Object),
            "metadata" => self.apply_metadata(operation).map(Value::Object),
            other => {
                let mut params = Map::new();
                params.insert("kind".into(), Value::from(other));
                Err(DomainError::new(
                    "agent.request_invalid",
                    "AgentRequestError",
                    "operation kind 非法",
                    params,
                )
                .into())
            }
        }
    }

    fn apply_edit(&self, operation: &OperationPlan) -> EngineResult<Map<String, Value>> {
        self.edit.apply(operation)
    }

    fn apply_migration(&self, operation: &OperationPlan) -> EngineResult<Map<String, Value>> {
        let params = operation.input()?;
        let source_tool = str_param(&params, "source_tool")?;
        let target_tool = str_param(&params, "target_tool")?;
        let reference = str_param(&params, "ref")?;
        let record = match self.index.resolve(&source_tool, &reference) {
            Ok(record) => record,
            Err(error) if error.error_type == "AgentReferenceError" => {
                return Err(DomainError::concurrent_modification(
                    "会话在迁移计划生成后已变化，请重新计划",
                )
                .into())
            }
            Err(error) => return Err(error.into()),
        };
        if record.revision != operation.base_revision {
            return Err(DomainError::concurrent_modification(
                "会话在迁移计划生成后已变化，请重新计划",
            )
            .into());
        }
        let session = self.index.read_indexed_session(&record)?;
        let result = self.migration.apply(
            &source_tool,
            &target_tool,
            session,
            None,
            params.get("max_turn").and_then(Value::as_i64),
            None,
        )?;
        // 二次门禁：回滚过、或结构验收不是恒等 true，都不许当成功收尾。
        let structure_ok = result
            .get("validation")
            .and_then(|validation| validation.get("structure"))
            .and_then(|structure| structure.get("ok"))
            == Some(&Value::Bool(true));
        if result.get("rolled_back").is_some_and(is_truthy) || !structure_ok {
            return Err(EngineError::runtime("迁移写入后的结构校验失败，产物已回滚"));
        }
        Ok(result)
    }

    fn apply_metadata(&self, operation: &OperationPlan) -> EngineResult<Map<String, Value>> {
        let params = operation.input()?;
        let tool = str_param(&params, "tool")?;
        let reference = str_param(&params, "ref")?;
        let session_id = str_param(&params, "session_id")?;
        let record = match self.index.resolve(&tool, &reference) {
            Ok(record) => record,
            Err(error) if error.error_type == "AgentReferenceError" => {
                return Err(DomainError::concurrent_modification(
                    "会话在元数据计划生成后已变化，请重新计划",
                )
                .into())
            }
            Err(error) => return Err(error.into()),
        };
        if record.revision != operation.base_revision {
            return Err(DomainError::concurrent_modification(
                "会话在元数据计划生成后已变化，请重新计划",
            )
            .into());
        }
        if record.row.get("id").and_then(Value::as_str) != Some(session_id.as_str()) {
            return Err(DomainError::concurrent_modification(
                "会话标识在元数据计划生成后已变化，请重新计划",
            )
            .into());
        }
        let expected = params
            .get("metadata_before")
            .and_then(Value::as_object)
            .cloned()
            .ok_or_else(|| EngineError::key_error("metadata_before"))?;
        let patch = params
            .get("patch")
            .and_then(Value::as_object)
            .cloned()
            .ok_or_else(|| EngineError::key_error("patch"))?;
        let applied =
            metadata::compare_and_set_entry(&tool, &session_id, &expected, &patch, &self.ports)?;
        let mut result = Map::new();
        result.insert("metadata".into(), Value::Object(applied));
        Ok(result)
    }
}

fn str_param(params: &Value, key: &str) -> EngineResult<String> {
    params
        .get(key)
        .and_then(Value::as_str)
        .map(str::to_string)
        .ok_or_else(|| EngineError::key_error(key))
}

/// Python 的真值判定：`result.get("rolled_back")` 只要是真值就算回滚过。
fn is_truthy(value: &Value) -> bool {
    match value {
        Value::Null => false,
        Value::Bool(flag) => *flag,
        Value::Number(number) => number.as_f64() != Some(0.0),
        Value::String(text) => !text.is_empty(),
        Value::Array(items) => !items.is_empty(),
        Value::Object(entries) => !entries.is_empty(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn truthiness_matches_python() {
        assert!(!is_truthy(&json!(null)));
        assert!(!is_truthy(&json!(false)));
        assert!(!is_truthy(&json!(0)));
        assert!(!is_truthy(&json!("")));
        assert!(!is_truthy(&json!([])));
        assert!(!is_truthy(&json!({})));
        assert!(is_truthy(&json!(true)));
        assert!(is_truthy(&json!("x")));
    }
}
